lib/assert.js

```diff
@@ -481,18 +479,30 @@ function partiallyCompareSets(actual, expected, comparedObjects) {
 
   if (isDeepEqual === undefined) lazyLoadComparison();
 
-  const actualArray = ArrayFrom(FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], actual));
+  // Create a map for faster lookups and counts
+  const actualMap = new SafeMap();
+  const actualIterator = FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], actual);
   const expectedIterator = FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], expected);
-  const usedIndices = new SafeSet();
 
-  expectedIteration: for (const expectedItem of expectedIterator) {
-    for (let actualIdx = 0; actualIdx < actualArray.length; actualIdx++) {
-      if (!usedIndices.has(actualIdx) && isDeepStrictEqual(actualArray[actualIdx], expectedItem)) {
-        usedIndices.add(actualIdx);
-        continue expectedIteration;
+  for (const actualItem of actualIterator) {
+    actualMap.set(actualItem, (actualMap.get(actualItem) || 0) + 1);
+  }
+
+  const actualMapIterator = FunctionPrototypeCall(SafeMap.prototype[SymbolIterator], actualMap);
+
+  for (const expectedItem of expectedIterator) {
+    let foundMatch = false;
+    for (const { 0: actualItem, 1: count } of actualMapIterator) {
```

Comment:
We only have to look for entries that are not primitives. That way we safe a lot of iterations. You can look at the code how it's down in the set comparison of `isDeepStrictEqual()`. It does exactly that. It can even be a bit simpler, because this does not require the non-strict checks.

Reply:
@BridgeAR very good point, I didn't think about it!